File: frontend/src/components/benchmark/content/a_star/implementations/a_star.js
class AStar {
    constructor() {
        this.graph = new Map();
    }

    addEdge(u, v) {
        if (!this.graph.has(u)) {
            this.graph.set(u, new Set());
        }
        if (!this.graph.has(v)) {
            this.graph.set(v, new Set());
        }
        this.graph.get(u).add(v);
        this.graph.get(v).add(u);
    }

    buildFromEdges(edges) {
        for (const [u, v] of edges) {
            this.addEdge(u, v);
        }
    }

    heuristic(node, goal) {
        return Math.abs(node - goal);
    }

    search(start, goal) {
        if (!this.graph.has(start) || !this.graph.has(goal)) {
            return null;
        }

        const openSet = new Map();
        const pq = [];
        const cameFrom = new Map();
        const gScore = new Map();
        const fScore = new Map();

        openSet.set(start, true);
        pq.push([start, 0]);
        cameFrom.set(start, start);
        gScore.set(start, 0);
        fScore.set(start, this.heuristic(start, goal));

        while (pq.length > 0) {
            pq.sort((a, b) => a[1] - b[1]);
            const current = pq.shift()[0];
            openSet.delete(current);

            if (current === goal) {
                const path = [];
                let node = current;
                while (node !== start) {
                    path.push(node);
                    node = cameFrom.get(node);
                }
                path.push(start);
                return path.reverse();
            }

            const neighbors = this.graph.get(current) || new Set();
            for (const neighbor of neighbors) {
                const tentativeG = gScore.get(current) + 1;

                if (!gScore.has(neighbor) || tentativeG < gScore.get(neighbor)) {
                    cameFrom.set(neighbor, current);
                    gScore.set(neighbor, tentativeG);
                    fScore.set(neighbor, tentativeG + this.heuristic(neighbor, goal));
                    
                    if (!openSet.has(neighbor)) {
                        openSet.set(neighbor, true);
                        pq.push([neighbor, fScore.get(neighbor)]);
                    }
                }
            }
        }

        return null;
    }
}

module.exports = AStar;

File: frontend/src/components/benchmark/content/a_star/implementations/a_star.rs
use std::collections::{HashMap, HashSet, BinaryHeap};
use std::cmp::Ordering;

#[derive(Copy, Clone, Eq, PartialEq)]
struct State {
    cost: f64,
    node: i32,
}

impl Ord for State {
    fn cmp(&self, other: &Self) -> Ordering {
        other.cost.partial_cmp(&self.cost).unwrap_or(Ordering::Equal)
    }
}

impl PartialOrd for State {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

pub struct AStar {
    graph: HashMap<i32, HashSet<i32>>,
}

impl AStar {
    pub fn new() -> Self {
        AStar {
            graph: HashMap::new(),
        }
    }

    pub fn add_edge(&mut self, u: i32, v: i32) {
        self.graph.entry(u).or_insert_with(HashSet::new).insert(v);
        self.graph.entry(v).or_insert_with(HashSet::new).insert(u);
    }

    pub fn build_from_edges(&mut self, edges: &[(i32, i32)]) {
        for &(u, v) in edges {
            self.add_edge(u, v);
        }
    }

    fn heuristic(&self, node: i32, goal: i32) -> f64 {
        (node - goal).abs() as f64
    }

    pub fn search(&self, start: i32, goal: i32) -> Option<Vec<i32>> {
        if !self.graph.contains_key(&start) || !self.graph.contains_key(&goal) {
            return None;
        }

        let mut open_set = BinaryHeap::new();
        let mut came_from = HashMap::new();
        let mut g_score = HashMap::new();
        let mut f_score = HashMap::new();

        open_set.push(State { cost: 0.0, node: start });
        came_from.insert(start, start);
        g_score.insert(start, 0.0);
        f_score.insert(start, self.heuristic(start, goal));

        while let Some(State { node: current, .. }) = open_set.pop() {
            if current == goal {
                let mut path = Vec::new();
                let mut node = current;
                while node != start {
                    path.push(node);
                    node = *came_from.get(&node)?;
                }
                path.push(start);
                path.reverse();
                return Some(path);
            }

            if let Some(neighbors) = self.graph.get(&current) {
                for &neighbor in neighbors {
                    let tentative_g = g_score.get(&current).unwrap_or(&f64::INFINITY) + 1.0;

                    if tentative_g < *g_score.get(&neighbor).unwrap_or(&f64::INFINITY) {
                        came_from.insert(neighbor, current);
                        g_score.insert(neighbor, tentative_g);
                        let f = tentative_g + self.heuristic(neighbor, goal);
                        f_score.insert(neighbor, f);
                        open_set.push(State { cost: f, node: neighbor });
                    }
                }
            }
        }

        None
    }
}